File: sdks/python/motosan_ai/providers/gemini_code_assist.py
from __future__ import annotations

import json
import time
from collections.abc import AsyncIterator
from dataclasses import dataclass, field
from itertools import count
from typing import Any

import httpx

from motosan_ai._stream_collect import collect_stream
from motosan_ai.error import AuthError, NetworkError, ProviderError, RateLimitError, StreamError
from motosan_ai.provider_base import BaseProvider, ProviderCapabilities
from motosan_ai.providers.gemini import build_gemini_body
from motosan_ai.retry import parse_retry_after_header
from motosan_ai.types import ChatRequest, ChatResponse, StopReason, StreamEvent, Usage

_DEFAULT_BASE_URL = "https://cloudcode-pa.googleapis.com"
_DEFAULT_MODEL = "gemini-2.5-flash"
_USER_AGENT = "google-cloud-sdk vscode_cloudshelleditor/0.1"
_X_GOOG_API_CLIENT = "gl-node/22.17.0"
_CLIENT_METADATA = (
    '{"ideType":"IDE_UNSPECIFIED","platform":"PLATFORM_UNSPECIFIED","pluginType":"GEMINI"}'
)
_REQUEST_COUNTER = count()
_TOOL_CALL_COUNTER = count()


def _http_error_kwargs(status: int, headers: httpx.Headers | None) -> dict[str, Any]:
    retry_after = None
    request_id = None
    if headers is not None:
        retry_after = parse_retry_after_header(headers.get("retry-after"))
        request_id = headers.get("request-id") or headers.get("x-request-id")
    return {"status_code": status, "retry_after": retry_after, "request_id": request_id}


def _gen_request_id() -> str:
    ts_ms = int(time.time() * 1000)
    return f"motosan-{ts_ms}-{next(_REQUEST_COUNTER):09d}"


def _gen_tool_call_id(name: str) -> str:
    ts_ms = int(time.time() * 1000)
    return f"{name}_{ts_ms}_{next(_TOOL_CALL_COUNTER)}"


@dataclass
class _CodeAssistAdapterState:
    seen_tool_ids: set[str] = field(default_factory=set)


_FINISH_REASON_MAP = {"MAX_TOKENS": StopReason.max_tokens}


def _parse_sse_event(data: str, state: _CodeAssistAdapterState) -> list[StreamEvent]:
    s = data.strip()
    if not s or s == "[DONE]":
        return []
    try:
        chunk = json.loads(s)
    except json.JSONDecodeError as exc:
        raise StreamError(f"malformed SSE chunk: {exc}") from exc

    response_data = chunk.get("response")
    if not isinstance(response_data, dict):
        return []

    candidates = response_data.get("candidates") or []
    if not candidates:
        return []
    candidate = candidates[0]
    if not isinstance(candidate, dict):
        return []

    parts = (candidate.get("content") or {}).get("parts") or []
    finish_reason = candidate.get("finishReason")
    out: list[StreamEvent] = []
    has_tool_calls = False

    for part in parts:
        if not isinstance(part, dict):
            continue
        text = part.get("text")
        if isinstance(text, str) and text:
            out.append(StreamEvent(content=text, done=False))
        fc = part.get("functionCall")
        if isinstance(fc, dict):
            has_tool_calls = True
            name = fc.get("name") or ""
            args = fc.get("args") or {}
            provided_id = fc.get("id")
            if (
                isinstance(provided_id, str)
                and provided_id
                and provided_id not in state.seen_tool_ids
            ):
                tool_id = provided_id
            else:
                tool_id = _gen_tool_call_id(name)
            state.seen_tool_ids.add(tool_id)
            out.append(
                StreamEvent(
                    content="",
                    done=False,
                    event_type="tool_call_start",
                    tool_call_id=tool_id,
                    tool_call_name=name,
                )
            )
            out.append(
                StreamEvent(
                    content="",
                    done=False,
                    event_type="tool_call_args",
                    tool_call_id=tool_id,
                    tool_call_args_delta=json.dumps(args),
                )
            )
            out.append(
                StreamEvent(
                    content="",
                    done=False,
                    event_type="tool_call_end",
                    tool_call_id=tool_id,
                )
            )

    usage_meta = response_data.get("usageMetadata")
    if isinstance(usage_meta, dict):
        prompt = int(usage_meta.get("promptTokenCount") or 0)
        cached = int(usage_meta.get("cachedContentTokenCount") or 0)
        output = int(usage_meta.get("candidatesTokenCount") or 0)
        out.append(
            StreamEvent(
                content="",
                done=False,
                event_type="usage",
                usage=Usage(
                    input_tokens=max(prompt - cached, 0),
                    output_tokens=output,
                    cache_read_input_tokens=cached if cached > 0 else None,
                ),
            )
        )

    if finish_reason:
        if has_tool_calls:
            stop = StopReason.tool_use
        elif finish_reason == "STOP":
            stop = StopReason.end_turn
        else:
            stop = _FINISH_REASON_MAP.get(finish_reason, StopReason.other)
        out.append(StreamEvent(content="", done=True, stop_reason=stop))

    return out


class GeminiCodeAssistProvider(BaseProvider):
    capabilities: ProviderCapabilities = ProviderCapabilities.with_image()

    def __init__(
        self,
        access_token: str,
        project_id: str,
        model: str | None = None,
        base_url: str | None = None,
    ) -> None:
        self.access_token = access_token
        self.project_id = project_id
        self.model = model or _DEFAULT_MODEL
        self.base_url = (base_url or _DEFAULT_BASE_URL).rstrip("/")
        self._http = httpx.AsyncClient(timeout=120.0)

    def _stream_url(self) -> str:
        return f"{self.base_url}/v1internal:streamGenerateContent?alt=sse"

    def _headers(self) -> dict[str, str]:
        return {
            "authorization": f"Bearer {self.access_token}",
            "user-agent": _USER_AGENT,
            "x-goog-api-client": _X_GOOG_API_CLIENT,
            "client-metadata": _CLIENT_METADATA,
            "content-type": "application/json",
        }

    def _build_envelope(self, request: ChatRequest) -> dict[str, Any]:
        inner = build_gemini_body(request)
        return {
            "project": self.project_id,
            "model": request.model or self.model,
            "request": inner,
            "userAgent": "motosan-ai",
            "requestId": _gen_request_id(),
        }

    @staticmethod
    def _map_http_error(
        status: int, message: str, headers: httpx.Headers | None = None
    ) -> Exception:
        metadata = _http_error_kwargs(status, headers)
        if status == 401:
            return AuthError(message, **metadata)
        if status == 429:
            return RateLimitError(message, **metadata)
        return ProviderError(message, **metadata)

    async def stream(self, request: ChatRequest) -> AsyncIterator[StreamEvent]:
        self.validate_request(request)
        body = self._build_envelope(request)
        try:
            resp = await self._http.send(
                self._http.build_request(
                    "POST", self._stream_url(), headers=self._headers(), json=body
                ),
                stream=True,
            )
        except httpx.HTTPError as exc:
            raise NetworkError(str(exc)) from exc

        try:
            if not resp.is_success:
                error_body = await resp.aread()
                raise self._map_http_error(resp.status_code, error_body.decode(), resp.headers)

            state = _CodeAssistAdapterState()
            try:
                async for line in resp.aiter_lines():
                    if not line.startswith("data: "):
                        continue
                    data = line[len("data: ") :]
                    for event in _parse_sse_event(data, state):
                        yield event
                        if event.done:
                            return
            except (StreamError, AuthError, RateLimitError, ProviderError, NetworkError):
                raise
            except httpx.HTTPError as exc:
                raise StreamError(f"stream transport error: {exc}") from exc
        finally:
            await resp.aclose()

    async def chat(self, request: ChatRequest) -> ChatResponse:
        response = await collect_stream(self.stream(request))
        return ChatResponse(
            content=response.content,
            tool_calls=response.tool_calls,
            model=request.model or self.model,
            usage=response.usage,
            stop_reason=response.stop_reason,
            thinking=response.thinking,
        )

File: sdks/rust/src/providers/gemini_code_assist.rs
use std::sync::atomic::{AtomicU64, Ordering};

use crate::error::MotosanError;
use crate::models::{DEFAULT_GEMINI_CODE_ASSIST_MODEL, GEMINI_CODE_ASSIST_BASE_URL};
use crate::providers::gemini::GeminiProvider;
use crate::providers::{
    extract_error_message, extract_request_id, map_http_error, parse_retry_after, send_with_retry,
    ProviderImpl,
};
use crate::retry::RetryPolicy;
use crate::stream::{collect_stream, BoxStream};
use crate::types::{ChatRequest, ChatResponse, StopReason, StreamEvent, Usage};
use async_trait::async_trait;
use eventsource_stream::Eventsource;
use futures_core::Stream;
use reqwest::Client;
use serde_json::{json, Value};
use std::collections::VecDeque;
use std::pin::Pin;
use std::task::Poll;
use std::time::{SystemTime, UNIX_EPOCH};

static REQUEST_COUNTER: AtomicU64 = AtomicU64::new(0);
static TOOL_CALL_COUNTER: AtomicU64 = AtomicU64::new(0);

fn gen_request_id() -> String {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis();
    let n = REQUEST_COUNTER.fetch_add(1, Ordering::Relaxed);
    format!("motosan-{ts}-{n:09}")
}

fn gen_tool_call_id(name: &str) -> String {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis();
    let n = TOOL_CALL_COUNTER.fetch_add(1, Ordering::Relaxed);
    format!("{name}_{ts}_{n}")
}

#[derive(Debug, Clone)]
pub struct GeminiCodeAssistProvider {
    http: Client,
    access_token: String,
    project_id: String,
    model: String,
    base_url: String,
    retry_policy: RetryPolicy,
}

impl GeminiCodeAssistProvider {
    pub fn new(
        access_token: impl Into<String>,
        project_id: impl Into<String>,
        model: Option<String>,
        base_url: Option<String>,
    ) -> Self {
        Self {
            http: Client::new(),
            access_token: access_token.into(),
            project_id: project_id.into(),
            model: model.unwrap_or_else(|| DEFAULT_GEMINI_CODE_ASSIST_MODEL.to_string()),
            base_url: base_url.unwrap_or_else(|| GEMINI_CODE_ASSIST_BASE_URL.to_string()),
            retry_policy: RetryPolicy::default(),
        }
    }

    pub fn with_retry_policy(mut self, policy: RetryPolicy) -> Self {
        self.retry_policy = policy;
        self
    }

    fn stream_url(&self) -> String {
        format!("{}/v1internal:streamGenerateContent?alt=sse", self.base_url)
    }

    fn build_envelope(&self, req: &ChatRequest) -> Value {
        let model = req.model.as_deref().unwrap_or(&self.model);
        // Reuse the public Gemini API request builder for the inner request fields
        let inner = GeminiProvider::build_request(req);
        json!({
            "project": self.project_id,
            "model": model,
            "request": inner,
            "userAgent": "motosan-ai",
            "requestId": gen_request_id(),
        })
    }

    fn apply_auth(&self, request: reqwest::RequestBuilder) -> reqwest::RequestBuilder {
        request
            .header("authorization", format!("Bearer {}", self.access_token))
            .header("user-agent", "google-cloud-sdk vscode_cloudshelleditor/0.1")
            .header("x-goog-api-client", "gl-node/22.17.0")
            .header(
                "client-metadata",
                r#"{"ideType":"IDE_UNSPECIFIED","platform":"PLATFORM_UNSPECIFIED","pluginType":"GEMINI"}"#,
            )
    }
}

#[async_trait]
impl ProviderImpl for GeminiCodeAssistProvider {
    fn capabilities(&self) -> crate::types::ProviderCapabilities {
        crate::types::ProviderCapabilities::with_image()
    }

    async fn chat(&self, req: ChatRequest) -> Result<ChatResponse, MotosanError> {
        let model = req.model.clone().unwrap_or_else(|| self.model.clone());
        let stream = self.stream(req).await?;
        let mut response = collect_stream(stream).await?;
        if response.model.is_empty() {
            response.model = model;
        }
        Ok(response)
    }

    async fn stream(&self, req: ChatRequest) -> Result<BoxStream, MotosanError> {
        let url = self.stream_url();
        let body = self.build_envelope(&req);

        let response = send_with_retry(&self.retry_policy, || {
            self.apply_auth(
                self.http
                    .post(&url)
                    .header("content-type", "application/json"),
            )
            .json(&body)
        })
        .await?;

        let status = response.status().as_u16();
        if status != 200 {
            let retry_after = parse_retry_after(response.headers());
            let request_id = extract_request_id(response.headers());
            let payload: Value = response.json().await.unwrap_or(json!({}));
            let msg = extract_error_message(&payload, "Gemini Code Assist error");
            return Err(map_http_error(status, msg, retry_after, request_id));
        }

        let sse = response.bytes_stream().eventsource();
        let adapter = CodeAssistStreamAdapter {
            inner: Box::pin(sse),
            pending: VecDeque::new(),
            seen_tool_ids: std::collections::HashSet::new(),
        };
        Ok(Box::pin(adapter))
    }
}

struct CodeAssistStreamAdapter {
    inner: Pin<
        Box<
            dyn Stream<
                    Item = Result<
                        eventsource_stream::Event,
                        eventsource_stream::EventStreamError<reqwest::Error>,
                    >,
                > + Send,
        >,
    >,
    pending: VecDeque<StreamEvent>,
    /// Track tool call IDs we've seen to detect duplicates (API may reuse IDs).
    seen_tool_ids: std::collections::HashSet<String>,
}

impl Stream for CodeAssistStreamAdapter {
    type Item = Result<StreamEvent, MotosanError>;

    fn poll_next(
        mut self: Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        if let Some(ev) = self.pending.pop_front() {
            return Poll::Ready(Some(Ok(ev)));
        }

        loop {
            match self.inner.as_mut().poll_next(cx) {
                Poll::Ready(Some(Ok(event))) => {
                    if event.data.trim() == "[DONE]" || event.data.trim().is_empty() {
                        continue;
                    }
                    let chunk: Value = match serde_json::from_str(&event.data) {
                        Ok(v) => v,
                        Err(_) => continue,
                    };

                    // Unwrap outer "response" field
                    let response_data = match chunk.get("response") {
                        Some(r) => r.clone(),
                        None => continue,
                    };

                    let candidate = match response_data.get("candidates").and_then(|c| c.get(0)) {
                        Some(c) => c.clone(),
                        None => continue,
                    };

                    let parts = candidate
                        .get("content")
                        .and_then(|c| c.get("parts"))
                        .and_then(Value::as_array)
                        .cloned()
                        .unwrap_or_default();

                    let finish_reason = candidate.get("finishReason").and_then(Value::as_str);

                    let mut has_tool_calls = false;

                    for part in &parts {
                        if let Some(text) = part.get("text").and_then(Value::as_str) {
                            if !text.is_empty() {
                                self.pending.push_back(StreamEvent::text(text));
                            }
                        }
                        if let Some(fc) = part.get("functionCall") {
                            has_tool_calls = true;
                            let name = fc
                                .get("name")
                                .and_then(Value::as_str)
                                .unwrap_or("")
                                .to_string();
                            let args = fc.get("args").cloned().unwrap_or(json!({}));
                            let args_str = args.to_string();

                            // Use API-provided ID if present and not a duplicate
                            let provided_id = fc
                                .get("id")
                                .and_then(Value::as_str)
                                .filter(|id| !id.is_empty());
                            let id = match provided_id {
                                Some(id) if !self.seen_tool_ids.contains(id) => id.to_string(),
                                _ => gen_tool_call_id(&name),
                            };
                            self.seen_tool_ids.insert(id.clone());

                            self.pending
                                .push_back(StreamEvent::tool_call_start(&id, &name));
                            self.pending
                                .push_back(StreamEvent::tool_call_args_with_id(&id, &args_str));
                            self.pending
                                .push_back(StreamEvent::tool_call_end_with_id(id));
                        }
                    }

                    // Usage
                    if let Some(usage_meta) = response_data.get("usageMetadata") {
                        let prompt_tokens = usage_meta
                            .get("promptTokenCount")
                            .and_then(Value::as_u64)
                            .unwrap_or(0) as u32;
                        let cached_tokens = usage_meta
                            .get("cachedContentTokenCount")
                            .and_then(Value::as_u64)
                            .unwrap_or(0) as u32;
                        let output_tokens = usage_meta
                            .get("candidatesTokenCount")
                            .and_then(Value::as_u64)
                            .unwrap_or(0) as u32;
                        // input_tokens excludes cached tokens (matching pi-mono behaviour)
                        let input_tokens = prompt_tokens.saturating_sub(cached_tokens);
                        self.pending.push_back(StreamEvent::usage(Usage {
                            input_tokens,
                            output_tokens,
                            cache_creation_input_tokens: None,
                            cache_read_input_tokens: if cached_tokens > 0 {
                                Some(cached_tokens)
                            } else {
                                None
                            },
                        }));
                    }

                    // Terminal event
                    if let Some(reason) = finish_reason {
                        let stop_reason = match reason {
                            "STOP" if has_tool_calls => StopReason::ToolUse,
                            "STOP" => StopReason::EndTurn,
                            "MAX_TOKENS" => StopReason::MaxTokens,
                            _ if has_tool_calls => StopReason::ToolUse,
                            _ => StopReason::Other,
                        };
                        self.pending
                            .push_back(StreamEvent::done_with_stop_reason(stop_reason));
                    }

                    if let Some(first) = self.pending.pop_front() {
                        return Poll::Ready(Some(Ok(first)));
                    }
                    continue;
                }
                Poll::Ready(Some(Err(e))) => {
                    return Poll::Ready(Some(Err(MotosanError::Stream(e.to_string()))));
                }
                Poll::Ready(None) => return Poll::Ready(None),
                Poll::Pending => return Poll::Pending,
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Message;

    fn provider() -> GeminiCodeAssistProvider {
        GeminiCodeAssistProvider::new("ya29.fake", "test-project", None, None)
    }

    #[test]
    fn capabilities_support_image_only() {
        let p = GeminiCodeAssistProvider::new("ya29.fake", "test-project", None, None);
        let caps = p.capabilities();
        assert!(caps.supports_image);
        assert!(!caps.supports_document);
    }

    #[test]
    fn stream_url_format() {
        let p = provider();
        assert!(p.stream_url().contains("cloudcode-pa.googleapis.com"));
        assert!(p.stream_url().contains("v1internal:streamGenerateContent"));
        assert!(p.stream_url().contains("alt=sse"));
    }

    #[test]
    fn envelope_wraps_inner_request() {
        let req = ChatRequest::builder()
            .messages(vec![Message::user("hello")])
            .system("be helpful")
            .build();
        let body = provider().build_envelope(&req);
        assert_eq!(body["project"], "test-project");
        assert_eq!(body["model"], DEFAULT_GEMINI_CODE_ASSIST_MODEL);
        assert_eq!(body["userAgent"], "motosan-ai");
        assert!(body["requestId"].as_str().unwrap().starts_with("motosan-"));
        // Inner request has contents and systemInstruction
        assert!(body["request"]["contents"].is_array());
        assert_eq!(
            body["request"]["systemInstruction"]["parts"][0]["text"],
            "be helpful"
        );
    }

    #[test]
    fn envelope_uses_request_model_override() {
        let req = ChatRequest::builder()
            .messages(vec![Message::user("hi")])
            .model("gemini-2.5-pro")
            .build();
        let body = provider().build_envelope(&req);
        assert_eq!(body["model"], "gemini-2.5-pro");
    }

    #[test]
    fn request_ids_are_unique() {
        let p = provider();
        let req = ChatRequest::builder()
            .messages(vec![Message::user("hi")])
            .build();
        let id1 = p.build_envelope(&req)["requestId"]
            .as_str()
            .unwrap()
            .to_string();
        let id2 = p.build_envelope(&req)["requestId"]
            .as_str()
            .unwrap()
            .to_string();
        assert_ne!(id1, id2);
    }

    mod stream_tests {
        use super::*;
        use futures_core::Stream;
        use tokio_stream::StreamExt;

        fn make_sse(
            data: &str,
        ) -> Pin<
            Box<
                dyn Stream<
                        Item = Result<
                            eventsource_stream::Event,
                            eventsource_stream::EventStreamError<reqwest::Error>,
                        >,
                    > + Send,
            >,
        > {
            let items = vec![Ok(eventsource_stream::Event {
                event: String::new(),
                data: data.to_string(),
                id: String::new(),
                retry: None,
            })];
            Box::pin(tokio_stream::iter(items))
        }

        #[tokio::test]
        async fn adapter_surfaces_inner_stream_error() {
            use eventsource_stream::EventStreamError;

            let utf8 = String::from_utf8(vec![0xff]).unwrap_err();
            let inner = tokio_stream::iter(vec![Err(EventStreamError::Utf8(utf8))]);
            let mut adapter = CodeAssistStreamAdapter {
                inner: Box::pin(inner),
                pending: VecDeque::new(),
                seen_tool_ids: std::collections::HashSet::new(),
            };

            let item = adapter.next().await.expect("one item");
            assert!(matches!(item, Err(MotosanError::Stream(_))));
        }

        #[tokio::test]
        async fn adapter_emits_text_from_response_wrapper() {
            let json = r#"{"response":{"candidates":[{"content":{"parts":[{"text":"hi!"}],"role":"model"},"finishReason":"STOP"}],"usageMetadata":{"promptTokenCount":3,"candidatesTokenCount":2}}}"#;
            let mut adapter = CodeAssistStreamAdapter {
                inner: make_sse(json),
                pending: VecDeque::new(),
                seen_tool_ids: std::collections::HashSet::new(),
            };
            let events: Vec<_> = (&mut adapter).collect::<Result<Vec<_>, _>>().await.unwrap();
            let text: String = events
                .iter()
                .filter(|e| e.event_type == crate::types::StreamEventType::Text)
                .map(|e| e.content.as_str())
                .collect();
            assert_eq!(text, "hi!");
            let done = events.iter().find(|e| e.done).expect("no done event");
            assert_eq!(done.stop_reason, Some(StopReason::EndTurn));
        }

        #[tokio::test]
        async fn adapter_uses_api_provided_tool_id() {
            let json = r#"{"response":{"candidates":[{"content":{"parts":[{"functionCall":{"name":"search","args":{"q":"rust"},"id":"api-provided-id"}}],"role":"model"},"finishReason":"STOP"}],"usageMetadata":{"promptTokenCount":5,"candidatesTokenCount":3}}}"#;
            let mut adapter = CodeAssistStreamAdapter {
                inner: make_sse(json),
                pending: VecDeque::new(),
                seen_tool_ids: std::collections::HashSet::new(),
            };
            let events: Vec<_> = (&mut adapter).collect::<Result<Vec<_>, _>>().await.unwrap();
            let start = events
                .iter()
                .find(|e| e.event_type == crate::types::StreamEventType::ToolCallStart)
                .unwrap();
            assert_eq!(start.tool_call_id.as_deref(), Some("api-provided-id"));
        }

        #[tokio::test]
        async fn adapter_generates_id_when_absent() {
            let json = r#"{"response":{"candidates":[{"content":{"parts":[{"functionCall":{"name":"calc","args":{}}}],"role":"model"},"finishReason":"STOP"}],"usageMetadata":{"promptTokenCount":2,"candidatesTokenCount":1}}}"#;
            let mut adapter = CodeAssistStreamAdapter {
                inner: make_sse(json),
                pending: VecDeque::new(),
                seen_tool_ids: std::collections::HashSet::new(),
            };
            let events: Vec<_> = (&mut adapter).collect::<Result<Vec<_>, _>>().await.unwrap();
            let start = events
                .iter()
                .find(|e| e.event_type == crate::types::StreamEventType::ToolCallStart)
                .unwrap();
            let id = start.tool_call_id.as_deref().unwrap_or("");
            assert!(!id.is_empty());
            assert!(id.starts_with("calc_"));
        }
    }
}
